src/MizarData/BaselineAndComparisonHelper.h

```diff
@@ -30,11 +32,92 @@ struct AddressToIdAndIdToSymbol {
       sfid_to_symbols;
 };
 
-[[nodiscard]] AddressToIdAndIdToSymbol AssignSampledFunctionIds(
-    const absl::flat_hash_map<orbit_mizar_base::AbsoluteAddress, orbit_mizar_base::FunctionSymbol>&
-        baseline_address_to_symbol,
-    const absl::flat_hash_map<orbit_mizar_base::AbsoluteAddress, orbit_mizar_base::FunctionSymbol>&
-        comparison_address_to_symbol);
+// `FunctionSymbolToKey` is default-constructible and defines `Key GetKey(const FunctionSymbol&)`
+// method. `Key` must be absl-hashable and define `operator==`. The functions with `FunctionSymbols`
+// mapped to equal `Key` will be assigned the same `SFID`.
+template <typename FunctionSymbolToKey>
```

Comment:
Are we planning to ever have something other than a string as key? I am wondering if we are over-engineering here.

Reply:
The decision to code to a generic `Key` has two positive implications
1. As you've suggested, we might have a non-string key. I am already anticipating keys like
```
struct Key {
std::string normalized_function_name;
std::string module_name;
bool is_exact_match; 
};
```

... or we could use... `using Key = Typedef<KeyTag, std::string>`.

2. The code doesn't rely on a generic-purpose `std::string`, but on a more descriptive Key, improving readability. 

What's for the code being over-engineered. "Over-engineering" suggests some drawbacks, yet I don't see any in this decision. E.g. the `BaselineAndComparisonHelper` would be templatized anyway, as we inject the `FunctionSymbolToKey`.

Reply:
>What's for the code being over-engineered. "Over-engineering" suggests some drawbacks, yet I don't see any in this decision.

Well, I would argue readability suffers.

> E.g. the `BaselineAndComparisonHelper` would be templatized anyway, as we inject the `FunctionSymbolToKey`.

This was my whole point. Why do you need the `FunctionSymbolToKey` as a template. Even the name suggest that the type is quite clear. You can either just pass a `map<FunctionSymbol, std::string>` (while I still think string as key suffices), or if you prefer you can have a `function<std::string(const FunctionSymbol&)>`.

Even if you really want to use something other than std::string as Key, you can still either have a virtual `FunctionSymbolKey` class, and derive from it, or an std::variant (if really needed).

My point is, that in general (more complex) templating should be the last resort and not the first choice. The problem that you are trying to solve, is a classical OOP problem. The solutions are know to everyone, so e.g. a reader would not need to wonder what `using Key = decltype(std::declval<FunctionSymbolToKey>().GetKey(std::declval<FunctionSymbol>()));` means.

Reply:
Also another benefit of having `FunctionSymbolToKey` and in particular `Key´ non-templated, is that when in an IDE, it is much easier to reason about the types, what functions get called, what do they mean, etc.

Reply:
> Even the name suggest that the type is quite clear. You can either just pass a `map<FunctionSymbol, std::string>` 

Even the dummy `FunctionSymbolToKey` I have for the next PR is more complex than just a map. It also has to check the modules.

>  if you prefer you can have a `function<std::string(const FunctionSymbol&)>`

Passing a lambda/function as `std::function` is usually a bad idea. At least, it does heap allocation and breaks inlining. The better way to pass a lambda is via templates. For that reason we've changed the `CallstackData::ForEach*` methods. 

> ... classical OOP problem ...

Yes. The problem could be solved via OOP instead via introducing one or two abstract classes. But I see no benefit to it. 

1. We have to design the interface in advance -- which does strike me as over-engineering. 
2. OOP solution would mean runtime DI, while all we need is compile-time DI. IMO when smth can be compile-time, it should be.
3. C++ standard library is `STL`, not `SOOPL`, so the concept of generic programming is just as "classical" as OOP is to C++.

About the `using Key` declaration. Everyone already knows what `decltype` is, I guess. I see this as rather a learning opportunity for the reader unfamiliar with `std::declval`. Leave alone, given the comment above the class, when the reader sees `using Key = `, it's already clear what the line is about. 

> Also another benefit of having FunctionSymbolToKey and in particular `Key´ non-templated

IMO it's pretty straightforward with `Key`, as we only use it as a key in a hash-map. I partially agree with you, that IDE support would be helpful in writing the code dealing with `GetKey` calls, but it helps writing, not reading and we should prioritise readability. Speaking of "what functions get called": if we went with an OOP solution, it wouldn't be any easier to figure out.

Reply:
`Everyone already knows what decltype is, I guess`

Wasn't true for me :).

Reply:
Then you've learned twice as much! :)

Reply:
>Even the dummy FunctionSymbolToKey I have for the next PR is more complex than just a map. It also has to check the modules.

Well then, a simple class would also do it.

>Yes. The problem could be solved via OOP instead via introducing one or two abstract classes. But I see no benefit to it.
> 1. We have to design the interface in advance -- which does strike me as over-engineering.

Why would you need to design the complete interface in advance. Start with what you know you need. We are free to change/adjust later. If you have two scenarios in mind, implement what unites them. In particular, you have the exact same problem with your templates. You know assume that there is a GetKey that is absl hashable, etc. You can just implement that same assumptions explicitly in a class/interface.

> 2. OOP solution would mean runtime DI, while all we need is compile-time DI. IMO when smth can be compile-time, it should be.

I would only agree in performance sensitive code. We are not in OrbitService here (and even there we sometimes rely on dynamic dispatch, we are also not in rendering or so). So I don't think is more important than readability (as you also say).

 >C++ standard library is STL, not SOOPL, so the concept of generic programming is just as "classical" as OOP is to C++.
 
 I agree with very simple templates. But the borderline to "complicated" templates is thing. And the style guide is quite clear here:
 ```
 Template metaprogramming sometimes allows cleaner and easier-to-use interfaces than would be possible without it, but it's also often a temptation to be overly clever. It's best used in a small number of low level components where the extra maintenance burden is spread out over a large number of uses.

Think twice before using template metaprogramming or other complicated template techniques; think about whether the average member of your team will be able to understand your code well enough to maintain it after you switch to another project, or whether a non-C++ programmer or someone casually browsing the code base will be able to understand the error messages or trace the flow of a function they want to call. If you're using recursive template instantiations or type lists or metafunctions or expression templates, or relying on SFINAE or on the sizeof trick for detecting function overload resolution, then there's a good chance you've gone too far.
 ```
 
 >About the using Key declaration. Everyone already knows what decltype is, I guess. I see this as rather a learning opportunity for the reader unfamiliar with std::declval. Leave alone, given the comment above the class, when the reader sees using Key = , it's already clear what the line is about.
 
Yes, as of the name, it is clear what this line is about. Still it is harder to read, then just having an abstract class Key.

>I partially agree with you, that IDE support would be helpful in writing the code dealing with GetKey calls, but it helps writing, not reading and we should prioritise readability.

But that's the point. Readability does get massively better, if your IDE has support for what you are doing. For your simple case, I see a call to `GetKey`. In non-templated code, I can control-click on it, get all implementations right away. See the documentation. With templates, I need to code search to find hopefully all definitions.

> Speaking of "what functions get called": if we went with an OOP solution, it wouldn't be any easier to figure out.

It does. Clion (and also VS) shows you all possible implementations. They also show you the virtual function declaration that might even contain a comment. In Clion, when your virtual function has a comment, you can even just hover over the call, and see that comment.